doc/api/v8.md

```diff
@@ -542,8 +542,248 @@ added: v8.0.0
 A subclass of [`Deserializer`][] corresponding to the format written by
 [`DefaultSerializer`][].
 
+## Promise hooks
+
+The `promise_hooks` interface can be used to track promise lifecycle events.
+To track _all_ async activity, see [`async_hooks`][] which internally uses this
+module to produce promise lifecycle events in addition to events for other
+async resources. For request context management, see [`AsyncLocalStorage`][].
+
+```mjs
+import { PromiseHooks } from 'v8';
+
+// There are four lifecycle events produced by promises:
+
+// The `init` event represents the creation of a promise. This could be a
+// direct creation such as with `new Promise(...)` or a continuation such
+// as `then()` or `catch()`. It also happens whenever an async function is
+// called or does an `await`. If a continuation promise is created, the
+// `parent` will be the promise it is a continuation from.
+function init(promise, parent) {
+  console.log('a promise was created', { promise, parent });
+}
+
+// The `resolve` event happens when a promise receives a resolution or
+// rejection value. This may happen synchronously such as when using
+// `Promise.resolve()` on non-promise input.
+function resolve(promise) {
+  console.log('a promise resolved or rejected', { promise });
+}
+
+// The `before` event runs immediately before a `then()` handler runs or
+// an `await` resumes execution.
+function before(promise) {
+  console.log('a promise is about to call a then handler', { promise });
+}
+
+// The `after` event runs immediately after a `then()` handler runs or when
+// an `await` begins after resuming from another.
+function after(promise) {
+  console.log('a promise is done calling a then handler', { promise });
+}
+
+// Lifecycle hooks may be started and stopped individually
+const stopWatchingInits = PromiseHooks.onInit(init);
+const stopWatchingResolves = PromiseHooks.onResolve(resolve);
+const stopWatchingBefores = PromiseHooks.onBefore(before);
+const stopWatchingAfters = PromiseHooks.onAfter(after);
+
+// Or they may be started and stopped in groups
+const stopAll = PromiseHooks.createHook({
+  init,
+  resolve,
+  before,
+  after
+});
+
+// To stop a hook, call the function returned at its creation.
+stopWatchingInits();
+stopWatchingResolves();
+stopWatchingBefores();
+stopWatchingAfters();
+stopAll();
+```
+
+### `PromiseHooks.onInit(init)`
+<!-- YAML
+added: REPLACEME
+-->
+
+* `init` {Function} The [`init` callback][] to call when a promise is created.
+* Returns: {Function} Call to stop the hook.
+
+```mjs
+import { PromiseHooks } from 'v8';
+
+const stop = PromiseHooks.onInit((promise, parent) => {});
+```
+
+```cjs
+const { PromiseHooks } = require('v8');
+
+const stop = PromiseHooks.onInit((promise, parent) => {});
+```
+
+### `PromiseHooks.onResolve(resolve)`
```

Comment:
Should we name it `onSettled` as it gets called for reject and resolve?

Reply:
I'm fine with that. 👍🏻